sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+ */
+typedef enum epmp_constants {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * Toggle for Rule Locking Bypass (mseccfg.RLB).
+ */
+typedef enum epmp_rlb_toggle {
+  kEpmpRlbToggleDisabled = 0, /**< mseccfg.RLB = 0 */
+  kEpmpRlbToggleEnabled,      /**< mseccfg.RLB = 1 */
+} epmp_rlb_toggle_t;
+
+/**
+ * EPMP state.
+ *
+ * A copy of a possible EPMP control register state.
+ */
+typedef struct epmp_state {
+  uint8_t pmpcfg[kEpmpNumRegions];
+  uintptr_t pmpaddr[kEpmpNumRegions];
+} epmp_state_t;
+
+/**
+ * EPMP entry permissions.
+ *
+ * Unlocked permissions can generally be modified when in M-mode. Locked
+ * permissions can only be modified in M-mode if Rule Locking Bypass
+ * (mseccfg.RLB) is set.
+ *
+ * When Machine Mode Lockdown is disabled (mseccfg.MML is unset) the
+ * combination R=0 W=1 is reserved. Note that this is the assumed state
+ * and so it is not possible to set these values.
+ *
+ * Note: these permissions have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ *
+ * TODO: MML mode could be supported by adding additional enums. That
+ * is why M-mode and S/U-mode permissions have been made explicit in
+ * enum names.
+ */
+typedef enum epmp_permissions {
+  // Unlocked permissions (can always be modified in machine mode).
+  // TODO: do we actually want to have unlocked rules? Could maybe remove
+  // these enums since MML will probably be enabled when in S/U-mode anyway.
+  kEpmpPermissionsUnlockedMachineAllUserNone = 0,    /**< LRWX = 0b0000 */
+  kEpmpPermissionsUnlockedMachineAllUserExecute,     /**< LRWX = 0b0001 */
+  kEpmpPermissionsUnlockedMachineAllUserRead,        /**< LRWX = 0b0100 */
+  kEpmpPermissionsUnlockedMachineAllUserReadExecute, /**< LRWX = 0b0101 */
+  kEpmpPermissionsUnlockedMachineAllUserReadWrite,   /**< LRWX = 0b0110 */
+  kEpmpPermissionsUnlockedMachineAllUserAll,         /**< LRWX = 0b0111 */
+
+  // Locked permissions (can only be modified in machine mode when mseccfg.RLB
+  // is set).
+  kEpmpPermissionsLockedMachineNoneUserNone,               /**< LRWX = 0b1000 */
+  kEpmpPermissionsLockedMachineExecuteUserExecute,         /**< LRWX = 0b1001 */
+  kEpmpPermissionsLockedMachineReadUserRead,               /**< LRWX = 0b1100 */
+  kEpmpPermissionsLockedMachineReadExecuteUserReadExecute, /**< LRWX = 0b1101 */
+  kEpmpPermissionsLockedMachineReadWriteUserReadWrite,     /**< LRWX = 0b1110 */
+  kEpmpPermissionsLockedMachineAllUserAll,                 /**< LRWX = 0b1111 */
+} epmp_permissions_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
+ * addresses).
+ *
+ * The intention is that this data structure is used to disambiguate the
+ * addresses of regions. It is therefore recommended that `start` and `end` are
+ * always used as labels when declaring a region.
```

Comment:
What does "declaring a region" mean? When I see "label" I think a label in an assembly file, so I think this prose could use some wordsmithing.

Reply:
Agreed it is badly worded. The recommendation is just that designated initializers should be used (i.e. `r = { .start = a, .end = b};` rather than `r = {a, b};`). It's probably redundant to say that though given the coding guidelines though so I think I'll just remove the comment.